frontend/providers/NativeAppProvider.tsx

```diff
@@ -0,0 +1,86 @@
+import React, {
+  useContext, useEffect, useMemo, useState,
+} from 'react';
+
+export type AppEvents = {
+  // This is the event that is fired when the native app has received a notification token
+  // and wants to send it to the website
+  SendNotificationToken: CustomEvent<{ token: string }>
+};
+
+export type NativeAppContext = {
+  isNativeApp: boolean
+  notificationToken?: string
+};
+
+const defaultContext: NativeAppContext = {
+  isNativeApp: false,
+};
+
+const nativeAppContext = React.createContext(defaultContext);
+
+export function NativeAppProvider({ children }) {
+  const [notificationToken, setNotificationToken] = useState(undefined);
+  useEffect(() => {
+    if (typeof window === 'undefined') {
+      return () => {};
+    }
+    // Use window object if notification token has been loaded before page loads
+    const initialNotificationToken = (window as any)?.notificationToken;
+    const localStorageToken = localStorage.getItem('notificationToken');
+    if (initialNotificationToken !== undefined) {
+      setNotificationToken(initialNotificationToken);
+    } else if (localStorageToken) {
+      setNotificationToken(localStorageToken);
+    }
+    // Use events if notification token loads, or updates, AFTER page initially loads
+    const onSendNotificationToken = (event: AppEvents['SendNotificationToken']) => {
+      setNotificationToken(event.detail.token);
+    };
+    window.addEventListener('appSendNotificationToken', onSendNotificationToken);
+    return () => {
+      window.removeEventListener('appSendNotificationToken', onSendNotificationToken);
+    };
+  }, []);
+
+  useEffect(() => {
+    localStorage.setItem('notificationToken', notificationToken ?? '');
```

Comment:
Good point, but in this case it will be stored in the local storage only if the user is visiting the web page from their mobile phone. In other words the notification token is stored on their mobile phone, encrypted behind default OS app data encryption. If someone could break this they could quite easily already get the user's ExpoToken if they wanted to.

However, it has now been a little while since I wrote this and I agree that it doesn't seem necessary to store the token locally, this can be removed.